src/d3a/models/market/market_structures.py

```diff
@@ -36,6 +36,16 @@ def __str__(self):
         return "{{{s.id!s:.6s}}} [{s.seller}]: {s.energy} kWh @ {s.price} @ {rate}"\
             .format(s=self, rate=self.price / self.energy)
 
+    def __hash__(self):
+        return hash(self.id)
+
+    def __eq__(self, other):
```

Comment:
Can you please tell me where this is used?

Reply:
Mostly in tests for now, but I thought it was a good addition in general. The root cause is the deepcopy call that creates a clone of the original offer object. The clone and the original offers are not the same objects, therefore the == comparison failed in the unit tests. In order to fix this behavior, the most straightforward way is to override the == operator (using the Python magic method __eq__) which defines the conditions on which an object is equal to another. This could of course be solved directly by changing the tests, but IMO this change is also nice to have to the production code too, because that way offer object comparisons will not have false negatives when comparing an offer with its clone, and that's the reason I added it globally.

Reply:
Ah, nice! Thanks for the info!